src/cljam/io/vcf/util/normalize.clj

```diff
@@ -0,0 +1,176 @@
+                      (into
+                       (map
+                        (fn [s]
+                          [s (biallelic-map (v s) format-id->number i n)]))
+                       samples))))))))))
+
+(defn same-ref?
+  "Checks if a REF allele matches the reference sequence."
+  [seq-reader {:keys [chr ^long pos ^String ref]}]
+  (let [ref-len (.length ref)
+        ref-region {:chr chr :start pos :end (dec (+ pos ref-len))}
+        ref-seq (io-seq/read-sequence seq-reader ref-region {:mask? false})]
+    (and ref-seq (.equalsIgnoreCase ^String ref-seq ref))))
+
+(defn- regions-before [chr ^long pos ^long window]
+  (->> (range pos 1 (- (dec window)))
+       (map
+        (fn [^long e]
+          {:chr chr,
+           :start (Math/max 1 (- e (dec window))),
+           :end (dec e)}))))
+
+(defn- char-equals-ignore-case
+  ([_]
+   true)
+  ([x y]
+   (= (Character/toUpperCase ^char x)
+      (Character/toUpperCase ^char y)))
+  ([x y & z]
+   (and (char-equals-ignore-case x y)
+        (apply char-equals-ignore-case y z))))
+
+(defn- trim-right
+  "Trims all duplicated allele sequences from right. If reached the start,
+  reference sequences of length `window` will be read from `seq-reader` and
+  be prepended to the alleles."
+  [seq-reader {:keys [chr ^long pos ref alt]
+               {:keys [END]} :info :as v} ^long window]
+  (let [alleles (cons ref alt)
+        min-end (long (dec (+ pos (apply min (map count alleles)))))
+        regs (regions-before chr pos window)
+        ref-seqs (sequence ;; unchunk
+                  (mapcat
+                   (fn [r]
+                     (-> seq-reader
+                         (io-seq/read-sequence r {:mask? false})
+                         cstr/reverse)))
+                  regs)
+        allele-seqs (map #(concat (cstr/reverse %) ref-seqs) alleles)
+        matched-length (->> allele-seqs
+                            (apply map char-equals-ignore-case)
+                            (take-while true?)
+                            count
+                            long
+                            (Math/min (dec min-end)))]
+    (if (pos? matched-length)
+      (let [pos' (or (some (fn [{:keys [start end]}]
+                             (when (<= start (- min-end matched-length) end)
+                               start)) regs)
+                     pos)
+            [ref' & alt'] (map
+                           #(->> %2
+                                 (take (inc (- (dec (+ pos (count %1))) pos')))
+                                 (drop matched-length)
+                                 reverse
+                                 cstr/join)
+                           alleles
+                           allele-seqs)]
+        (-> (if (integer? END)
+              (assoc-in v [:info :END] (dec (+ pos' (count ref'))))
+              v)
+            (assoc :pos pos')
+            (assoc :ref ref')
+            (assoc :alt alt')))
```

Comment:
`(assoc map key val)` is also `(assoc map key val & kvs)`, so it can be written as:

```clojure
(assoc :pos pos' :ref ref' :alt alt')
```